In [394]:
import numpy as np # imported to calculate euclidian distance between multidimensional arrays.
from sklearn.datasets import load_iris # imported to load iris dataset
from sklearn.model_selection import train_test_split # imported to split dataset (%70train,%30test)
from collections import Counter # imported to find out amount of species which are in the neighbor list.
from sklearn.metrics import accuracy_score # imported to compare with my predicts.

In [395]:
ds_iris= load_iris() #loaded iris data set

In [396]:
data=ds_iris.data # split data and target. 
species= ds_iris.target 

In [397]:
data_train,data_test,species_train,species_test = train_test_split(data,species,test_size=0.3) #split data set as train and test
 (%70train,%30test)

In [398]:
data_train[1] #trying to find out how is look like my data or row.

array([5.8, 2.7, 4.1, 1. ])

In [399]:
# implemention to fit data
def my_fit(data_train,species_train):    
    data_train=data_train
    species_train=species_train

In [400]:
# calculation of euclidian distance.
# i used numpy to make calculations beacuse on this dataset we have arrays.
# sum, sqrt and square functions are used to find distance between multidimensional arrays.
# the formula between the two point((x, y), (a, b)) is  = √((x - a)² + (y - b)²)
def euc_distance(a,b):
    result1=np.sum(np.square(a-b))
    result = np.sqrt(result1)
    return result
    

In [401]:
#checked if my method working or not.
a=np.array([3,0])
b=np.array([0,4])

In [402]:
print(euc_distance(a,b))

5.0


In [403]:
#my first goal here is to find neighbors of my new data or point.
#
def my_neighbor(data,k_value):
    
    neighbors= [] # list for neighbors
    distances=[] # list to store distances between points and data.
    for i in range(0,len(data_train)):   # this for loop taking the new point and comparing with all points and data we have
                                                                                                                #on train set.
        distance = euc_distance(data,data_train[i]) # It compares and finds distances between them.
        distances.append((species_train[i],distance)) # add the list with their target info and distances.
                                                    # target info: species_train[i] it gives the target like 0,1,2
                                                    # (Iris setosa, Iris virginica and Iris versicolor)
                                                     # it gives the species. 

    distances.sort(key=lambda distance: distance[1]) # sorting the list according to distances 
                                                     #(distance[0]:targets, distance[1]:distances)
                                                # i sorted the list beacuse i need the nearest neighbors.
                                                # the lowest distance means that nearest neighbor.
    for i in range(k_value):
        neighbors.append(distances[i][0])  #it takes targets from distances list and adding to neighbor list 
                                            # it takes as much as k value. (k value= amount of neighbors we want to check)
        
    times=Counter(neighbors) # it checks how much target we have. (how much target0 we have, target1 we have etc.)
                            # targets gives us species of flower. (0, 1 or 2) Iris setosa, Iris virginica and Iris versicolor
                        
    return max(times) # it returns the number of target (we have the most) 
                    # if k=7 and we have 3 target0, 2 target1 and 2 target2 it returns 0. (Iris setosa)
    
    

In [404]:
#note:I checked Counter method and its gives as the result as a dictionary type.
# for example: (k=5)Counter({2: 4, 1: 1}) means that we have 4 target2 and 1 target1. 
#after that i tried max method. it gave to me 2. then i implemented this counter function to my method. (my_neighbor)

In [405]:
#testing the method. and it gives the right result.
dst= my_neighbor(data_test,3)
dst

1

In [406]:
def my_predict(data_test,k_value): #taking data and k value.
    predict_list= []         #list to store all results. we will compare with real results and find out our accuracyscore later
    for data in data_test:    #this for loop find out the all neighbors of all points we have on dataset(testset%30).
        species1=my_neighbor(data,k_value) #using our previous neighbor method.
        predict_list.append(species1) # adding our predicts to the list. 
    return predict_list # this list should look like : [2,2,1,0,1,1,0,2 ...etc.] our findings are our targets.(species)


In [407]:
my_fit(data_train,species_train)

In [408]:
result_test1= my_predict(data_test,5) # checking method for k=5

In [409]:
accuracy_score(species_test,result_test1) # checking our accuracy score.

0.9555555555555556

In [410]:
for i in range(1,16,2): #checking our accuracy score for k=1,k=3,k=5,k=7,k=9,k=11,k=13,k=15 
    result_test= my_predict(data_test,i)  # i used odd numbers for k because i want to get a more stable results.
                                          # even numbers might be a problem in case of equality situtaions 
                                                                    # like (k=6 target0=2,target1=2,target2=2)
    print("k="+ str(i) + " :" + str(accuracy_score(species_test,result_test)) )

k=1 :0.9555555555555556
k=3 :0.9777777777777777
k=5 :0.9555555555555556
k=7 :0.9333333333333333
k=9 :0.8888888888888888
k=11 :0.8888888888888888
k=13 :0.8666666666666667
k=15 :0.8666666666666667


# My findings according to different k values:
 #k=1 :0.9333333333333333
 #k=3 :0.9555555555555556
 #k=5 :0.9555555555555556
 #k=7 :0.9333333333333333
 #k=9 :0.9555555555555556
 #k=11 :0.9111111111111111
 #k=13 :0.9111111111111111
 #k=15 :0.8888888888888888

# summary
I tried many times and mostly i find out k=3 and k 5 is giving the most decent results.
When i try to bigger numbers, generally accuracy score  decreased.
It shows us, knn algorithm is can not used for all situtaions. We can not trust the accuracy of knn algorithm.